### PR TITLE
Quick enter button

### PR DESCRIPTION
**Pull request by loic-martinez-freelance**
Added a quick enter button on the giveaway list. (With a setting to
enable / disable it)

### CODE REVIEW

**Comment by loic-martinez-freelance** [comment]
Added a screenshot

File: core/src/main/java/net/mabako/steamgifts/adapters/viewholder/GiveawayListItemViewHolder.java [hunk 1]
```diff
@@ -136,6 +140,35 @@ public void onError() {
...
+            giveawayEnterButton.setVisibility(View.GONE);
+        } else {
+            // Set the correct text based on the giveaway is already entered or not
+            if (giveaway.isEntered()) {
+                giveawayEnterButton.setText(activity.getString(R.string.leave_giveaway));
+                giveawayEnterButton.setEnabled(true);
+            } else {
+                giveawayEnterButton.setText(activity.getString(R.string.enter_giveaway));
+                //Check if the user have enough points, that the giveaway is not from him and that he have the required level
+                giveawayEnterButton.setEnabled(giveaway.getPoints() <= SteamGiftsUserData.getCurrent(null).getPoints() && giveaway.getLevel() <= SteamGiftsUserData.getCurrent(null).getLevel() && !SteamGiftsUserData.getCurrent(null).getName().equals(giveaway.getCreator()));
```
Comment (mabako) [inline core/src/main/java/net/mabako/steamgifts/adapters/viewholder/GiveawayListItemViewHolder.java]: How does the context menu determine that you can enter a giveaway? This looks like it should be a shared function.

File: core/src/main/java/net/mabako/steamgifts/adapters/viewholder/GiveawayListItemViewHolder.java [hunk 2]
```diff
@@ -136,6 +140,35 @@ public void onError() {
...
+
+        // Initialize the enter button
+        // Check if logged or the quick enter button setting is enabled
+        boolean loggedIn = SteamGiftsUserData.getCurrent(null).isLoggedIn();
+        if (!loggedIn || !PreferenceManager.getDefaultSharedPreferences(fragment.getContext()).getBoolean("preference_giveaway_show_quick_enter", false)){
+            giveawayEnterButton.setVisibility(View.GONE);
+        } else {
+            // Set the correct text based on the giveaway is already entered or not
+            if (giveaway.isEntered()) {
+                giveawayEnterButton.setText(activity.getString(R.string.leave_giveaway));
```
Comment (mabako) [inline core/src/main/java/net/mabako/steamgifts/adapters/viewholder/GiveawayListItemViewHolder.java]: I would prefer if this was an icon from font-awesome. The list of hidden games is an example where a "x" is used for the quick action, I think, which I feel is more appropriate. 

This issue likely is worse on low resolution devices.

File: core/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -256,4 +256,7 @@
     <!-- unsorted -->
     <string name="could_not_load_image">Could not load image.</string>
     <string name="share_image">Share Image</string>
+    <string name="preference_giveaway_quick_enter">Display quick enter button</string>
```
Comment (mabako) [inline core/src/main/res/values/strings.xml]: Note to self: check for consistency when I'm home

File: core/src/main/java/net/mabako/steamgifts/adapters/viewholder/GiveawayListItemViewHolder.java [hunk 4]
```diff
@@ -136,6 +140,35 @@ public void onError() {
         indicatorLevelNegative.setVisibility(giveaway.isLevelNegative() ? View.VISIBLE : View.GONE);
         indicatorPrivate.setVisibility(giveaway.isPrivate() ? View.VISIBLE : View.GONE);
         indicatorRegionRestricted.setVisibility(giveaway.isRegionRestricted() ? View.VISIBLE : View.GONE);
+
+        // Initialize the enter button
+        // Check if logged or the quick enter button setting is enabled
+        boolean loggedIn = SteamGiftsUserData.getCurrent(null).isLoggedIn();
+        if (!loggedIn || !PreferenceManager.getDefaultSharedPreferences(fragment.getContext()).getBoolean("preference_giveaway_show_quick_enter", false)){
```
Comment (mabako) [inline core/src/main/java/net/mabako/steamgifts/adapters/viewholder/GiveawayListItemViewHolder.java]: The default value in your settings definition is true, but is assumed to be false here.

File: core/src/main/java/net/mabako/steamgifts/fragments/GiveawayListFragment.java [hunk 5]
```diff
@@ -200,7 +200,8 @@ public void onEnterLeaveResult(String giveawayId, String what, Boolean success,
                     // we want to hide entered giveaways
                     adapter.removeGiveaway(giveawayId);
                 } else {
-                    adapter.notifyItemChanged(giveaway);
+                    // We refresh the entire dataset to let the other quick enter buttons know if the user still have enough points
+                    adapter.notifyDataSetChanged();
```
Comment (mabako) [inline core/src/main/java/net/mabako/steamgifts/fragments/GiveawayListFragment.java:204]: What are the performance implications for this?  Would it make sense to only refresh all items if the setting is enabled?

**Review by mabako — COMMENTED** [review]
I'll have a more in-depth look at it when I'm back home (probably tomorrow), but apart from disliking the button appearance it looks good.

**Comment by mabako** [comment]
Thanks for your pull request!

**Pull request merged**